bufferevent_mbedtls.c

```diff
@@ -0,0 +1,1379 @@
+
+		printf("%s in %s %s\n", msg, lib, func);
+	}
+}
+#else
+#define print_err(v) ((void)0)
+#endif
+
+
+/* Called to extract data from the BIO. */
+static int
+bio_bufferevent_read(void *ctx, unsigned char *out, size_t outlen)
+{
+	struct bufferevent *bufev = (struct bufferevent*)ctx;
+	int r = 0;
+	struct evbuffer *input;
+
+	//BIO_clear_retry_flags(b);
+	fprintf(stdout, "bio prepare write:\n");
+	fwrite(out, 1, outlen, stdout);
+
+	if (!out)
+		return 0;
+	if (!bufev)
+		return MBEDTLS_ERR_NET_INVALID_CONTEXT;
+
+	input = bufferevent_get_input(bufev);
+	if (evbuffer_get_length(input) == 0) {
+		/* If there's no data to read, say so. */
+		//BIO_set_retry_read(b);
+		return MBEDTLS_ERR_SSL_WANT_READ;
+	} else {
+		r = evbuffer_remove(input, out, outlen);
+	}
+	fprintf(stderr, "bio read %d bytes\n", r);
+
+	return r;
+}
+
+/* Called to write data info the BIO */
+static int
+bio_bufferevent_write(void *ctx, const unsigned char *in, size_t inlen)
+{
+	struct bufferevent *bufev = (struct bufferevent*)ctx;
+	struct evbuffer *output;
+	size_t outlen;
+
+	//BIO_clear_retry_flags(b);
+
+	if (!bufev)
+		return MBEDTLS_ERR_NET_INVALID_CONTEXT;
+
+	output = bufferevent_get_output(bufev);
+	outlen = evbuffer_get_length(output);
+
+	/* Copy only as much data onto the output buffer as can fit under the
+	 * high-water mark. */
+	if (bufev->wm_write.high && bufev->wm_write.high <= (outlen+inlen)) {
+		if (bufev->wm_write.high <= outlen) {
+			/* If no data can fit, we'll need to retry later. */
+			//BIO_set_retry_write(b);
+			return MBEDTLS_ERR_SSL_WANT_WRITE;
+		}
+		inlen = bufev->wm_write.high - outlen;
+	}
+
+	EVUTIL_ASSERT(inlen > 0);
+	evbuffer_add(output, in, inlen);
+	fprintf(stderr, "bio write %zu bytes\n", inlen);
+	return inlen;
+}
+
+
+/* --------------------
+   Now, here's the OpenSSL-based implementation of bufferevent.
+
+   The implementation comes in two flavors: one that connects its SSL object
+   to an underlying bufferevent using a BIO_bufferevent, and one that has the
+   SSL object connect to a socket directly.  The latter should generally be
+   faster, except on Windows, where your best bet is using a
+   bufferevent_async.
+
+   (OpenSSL supports many other BIO types, too.  But we can't use any unless
+   we have a good way to get notified when they become readable/writable.)
+   -------------------- */
+
+struct bio_data_counts {
+	unsigned long n_written;
+	unsigned long n_read;
+};
+
+struct bufferevent_mbedtls {
+	/* Shared fields with common bufferevent implementation code.
+	   If we were set up with an underlying bufferevent, we use the
+	   events here as timers only.  If we have an SSL, then we use
+	   the events as socket events.
+	 */
+	struct bufferevent_private bev;
+	/* An underlying bufferevent that we're directing our output to.
+	   If it's NULL, then we're connected to an fd, not an evbuffer. */
+	struct bufferevent *underlying;
+    /* net fd */
+    mbedtls_net_context net_ctx;
```

Comment:
these two lines uses spaces to indent, while the surrounding lines use tabs to indent